src/schedule/zephyr_ll.c

```diff
@@ -444,7 +426,7 @@ static int zephyr_ll_task_cancel(void *data, struct task *task)
 	 */
 	zephyr_ll_lock(sch, &flags);
 	if (task->state != SOF_TASK_STATE_FREE)
-		task->state = SOF_TASK_STATE_CANCEL;
+		zephyr_ll_task_done(sch, task);
```

Comment:
Btw, I think we probably have a chance after we align on this PR to simplify some of the scheduler APIs for Zephyr usage (since we will have a good understanding of what we need to support timers, dma as scheduling triggers).

Reply:
@lyakh `domain_register()`/`domain_unregister()` is called for every task, otherwise DMA domain wouldn't even have a chance of working since `domain_register()` is the one enabling the DMA IRQs (I can provide screenshots of the logs showing `domain_unregister()`/`domain_register()` being called multiple times if need be). Since we can have multiple DMA IRQs `domain_register()` should and is called for each task. What makes you say it's only called for first/last tasks? Also, that's the only logical explanation I could find for the following behavior:

1) Initially, when a playback and a capture are scheduled the capture calls  `domain_unregister()` when finished and works fine but playback doesn't call `domain_unregister()` so, scheduling another playback immediately after this doesn't work.
2) Trying to schedule a playback immediately after another playback doesn't work since the first playback doesn't call `domain_unregister()`.
3) Trying to schedule a capture after a playback works fine since they use different channels so `zephyr_ll_run` is still able to be executed thanks to the fact that capture's DMA IRQs are still enabled.